add SIP for governance.md (#2593)
fermyon/spin: docs/content/sips/019-governance.md

title = "SIP 019 - Spin Governance"
template = "main"
date = "2024-06-18T23:00:00Z"

---

Summary: Adding governance to the Spin project

Owner: michelle@fermyon.com

Created: June 18, 2024

## Background

Spin is a project now used by many organizations and individuals across the software ecosystem. In order for the Spin community to continue to rely on, build on, and integrate with Spin, there should be a public collective understanding of how the Spin project is governed and how to iterate on governance.

## Proposal

> This SIP is a proposal to add the following in a Governance.md file to the root of the Spin repository. This governance structure aims to describe transparently how we've been doing things so far. It formally opens up the project for maintainers from multiple organizations. It also includes some future directives for working toward a proper multi stakeholder governed project.

Reviewer: I think we should expand this a bit to talk about how this governance model is explicitly temporary and meant as a bootstrapping process. In the governance document we say the following:

> In addition to the responsibilities listed above, [the Spin Governance Committee] is also responsible for bootstrapping a multi-stakeholder steering committee of 5-7 people to govern the project. The group is responsible for creating a steering committee Charter and shaping and executing the processes around selecting committee members. Once a steering committee is in place, the Spin Governance Committee will be dismantled and replaced with the Spin Steering Committee. The Spin Steering Committee will be responsible for bootstrapping a Code of Conduct Committee.

I think we're burying the lead by not stating the explicit goal of progressing the status quo to a new governance structure. Something like this could be added here in the introduction:

> The current structure outlined in the Governance.md document is established as a temporary bootstrapping phase on the road to a more permanent, multi-stakeholder model. The Spin Governance Committee established in this document should eventually be dismantled and replaced with a smaller, more well defined group only tasked with governing the Spin project as a whole (a "Steering Committee"). This group will in turn establish other committees for handling more specific governance concerns (e.g., Project moderation and CoC enforcement).

> ```Governance.md

Reviewer: I think this can be removed now that it's in the dedicated governance.md file, right?

Project maintainers do not need to be software developers. No explicit role is placed upon them, and they can be anyone appropriate for the work being produced.

Project maintainers and emeritus maintainers for each project should be outlined in the MAINTAINERS.md file in the corresponding GitHub repository.

Changes to project maintainers use the following:

- A project maintainer may step down by emailing the mailing list. When a project maintainer steps down, they become an emeritus maintainer.
- Project maintainers MUST remain active on the project. If they are unresponsive for > 3 months, they will lose project maintainer-ship, unless a [supermajority](https://en.wikipedia.org/wiki/Supermajority#Two-thirds_vote) (two-thirds majority) of the project maintainers of the given project agrees to extend the period to be greater than 3 months. If two-thirds majority is not possible, the Spin Governance Committee members will be included in the group of people voting on this decision in addition to the project maintainers of the project.

> - New maintainers can be added to a project by a [supermajority](https://en.wikipedia.org/wiki/Supermajority#Two-thirds_vote) vote of the project’s existing maintainers. New maintainers must be nominated by existing maintainers. If quorum is not met, the Spin Govenance Committee members will be included in this vote.

Author: I think my only reason for this is that voting and specifically voting in private allows current maintainers to raise concerns more readily.

> - New maintainers can be added to a project by a [supermajority](https://en.wikipedia.org/wiki/Supermajority#Two-thirds_vote) vote of the project’s existing maintainers. New maintainers must be nominated by existing maintainers. If quorum is not met, the Spin Govenance Committee members will be included in this vote.

Reviewer: The point about keeping the conversation private makes sense though we should explicitly note that in the text. However, I don't see objection-free consensus and privacy as being mutually exclusive. Objection-free consensus is about ensuring that there isn't a culture of decision making where there are winners and losers - if possible everyone should feel at worst that their concerns were heard and not blocking objections exist.

## **Spin Governance Committee**

The project maintainers for [github.com/fermyon/spin](http://github.com/fermyon/spin) also serve as the Spin Governance Committee and have the following additional responsibilities:

- Maintaining the mission, vision, values, and scope of the project
- Refining this governance document
- Making project level decisions
- Making decisions about project licensing, contribution requirements, and other legal matters
- Resolving escalated project decisions when responsible project maintainers are blocked
- Managing the Spin brand
- Managing access to Spin assets such as source repositories, hosting, project calendars
- Handling code of conduct violations for any repository under the Spin project
- Deciding what sub-groups and repositories are part of the Spin project
- Overseeing the resolution and disclosure of security issues
- Managing financial decisions related to the project

> In addition to the responsibilities listed above, this group is also responsible for bootstrapping a multi-stakeholder steering committee of 5-7 people to govern the project. The group is responsible for creating a steering committee Charter and shaping and executing the processes around selecting committee members. Once a steering committee is in place, the Spin Governance Committee will be dismantled and replaced with the Spin Steering Committee. The Spin Steering Committee will be responsible for bootstrapping a Code of Conduct Committee.

Reviewer: It'd be nice to be a wee bit less perspective as to exactly what the steering committee has to do and instead nudge it in the right direction while still explicitly giving it the freedom to do things as it sees fit. 

```suggestion
In addition to the responsibilities listed above, this group is also responsible for bootstrapping a multi-stakeholder steering committee of 5-7 people to govern the project. The group is responsible for creating a steering committee Charter and shaping and executing the processes around selecting committee members. Once a steering committee is in place, the Spin Governance Committee will be dismantled and replaced with the Spin Steering Committee. The Spin Steering Committee should then establish additional governance structures as it sees fit (e.g., a Code of Conduct Committee for project moderation).
```

The Spin Governance Committee have to match the following criteria:

> - Spin Governance Committee MUST remain active on the project. If they are unresponsive for > 3 months, they will lose membership, unless a [supermajority](https://en.wikipedia.org/wiki/Supermajority#Two-thirds_vote) of the other governance members agrees to extend the period to be greater than 3 months

Reviewer: I believe this is what was meant (right now it sort of reads like we're talking about the committee as a whole instead of requirements placed on the individual members). 

```suggestion
- Spin Governance Committee members MUST remain active on the project. If they are unresponsive for > 3 months, they will lose membership, unless a [supermajority](https://en.wikipedia.org/wiki/Supermajority#Two-thirds_vote) of the other governance members agrees to extend the period to be greater than 3 months
```

> The Spin Governance Committee will select a chair to set agendas, call meetings, and oversee voting.

Reviewer: Nit: voting is only a part of the decision making process:
```suggestion
The Spin Governance Committee will select a chair to set agendas, call meetings, and oversee the decision making process.
```

### **Decision Making**

The default decision making process is objection-free consensus except for changes to the Governance of the project which requires a supermajority agreement of the committee.

Decision making scenarios should be promoted appropriately by the member overseeing the issue.

When a consensus cannot be found, a member can call for a supermajority vote from the committee.

### Changes to Governance.md

The process for changes to the Governance.md file is as follows:

1) Open an issue

> 2) Open a pull request to this SIP describing changes and the Governance.md file with proposed changes. 

Reviewer: It feels strange that future changes would not require a SIP. I could understand small clarifications that don't obviously materially change the intent of the governance.md doc being done through PR, but IMO any changes of substance should go through the SIP process.

Author: That makes sense to me. Thanks @tschneidereit. Updated with this process.

title = "SIP 019 - Spin Governance"
template = "main"
date = "2024-06-18T23:00:00Z"

---

Summary: Adding governance to the Spin project

Owner: michelle@fermyon.com

Created: June 18, 2024

## Background

Spin is a project now used by many organizations and individuals across the software ecosystem. In order for the Spin community to continue to rely on, build on, and integrate with Spin, there should be a public collective understanding of how the Spin project is governed and how to iterate on governance.

## Proposal

> This SIP is a proposal to add a GOVERNANCE.md file to the root of the Spin repository. The governance structure described in the corresponding GOVERNANCE.md file aims to describe transparently a process close to how we've been doing things so far in the project and how to make changes. It formally opens up the project for maintainers from multiple organizations, defines maintainership roles, describes decision making processes and more.

Reviewer: We should try to remember to link to the current version of `GOVERNANCE.md` so that it's easy in the future to see the exact version of the governance.md that was agreed to by this SIP (without having to crawl GitHub). Of course, we can't link to the exact version until we're sure it's final.
